src/Data/Text/Utf8/AhoCorasick/Automaton.hs

```diff
@@ -49,47 +49,49 @@ import Data.Word (Word32, Word64)
 import qualified Data.Char as Char
 import qualified Data.IntMap.Strict as IntMap
 import qualified Data.List as List
+import Data.Primitive (Prim)
 import qualified Data.Vector as Vector
-import qualified Data.Vector.Unboxed as UVector
 
 import Data.Text.CaseSensitivity (CaseSensitivity (..))
 import Data.Text.Utf8 (CodeUnit, CodeUnitIndex (CodeUnitIndex), Text (..), indexTextArray)
+import Data.TypedByteArray (TypedByteArray)
 
 import qualified Data.Text.Utf8 as Utf8
+import qualified Data.TypedByteArray as TBA
 
 -- TYPES
 -- | A numbered state in the Aho-Corasick automaton.
 type State = Int
 
--- | A transition is a pair of (code unit, next state). The code unit is 8 bits,
--- and the state index is 32 bits. We pack these together as a manually unlifted
--- tuple, because an unboxed Vector of tuples is a tuple of vectors, but we want
--- the elements of the tuple to be adjacent in memory. (The Word64 still needs
--- to be unpacked in the places where it is used.) The code unit is stored in
--- the least significant 32 bits, with the special value 2^8 indicating a
--- wildcard; the "failure" transition. Bit 9 through 31 (starting from zero,
+-- | A transition is a pair of (code point, next state). The code point is 21 bits,
+-- and the state index is 32 bits. The code point is stored in
+-- the least significant 32 bits, with the special value 2^21 indicating a
+-- wildcard; the "failure" transition. Bits 22 through 31 (starting from zero,
 -- both bounds inclusive) are always 0.
 --
 --
 -- >  Bit 63 (most significant)                 Bit 0 (least significant)
 -- >  |                                                                 |
 -- >  v                                                                 v
--- > |<--       goto state         -->|<--       zeros     -->| |<-input>|
--- > |SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS|00000000000000000000000|W|IIIIIIII|
--- >                                                           |
--- >                                                 Wildcard bit (bit 8)
+-- > |<--       goto state         -->|<-- 0s -->| |<--     input     -->|
+-- > |SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS|0000000000|W|IIIIIIIIIIIIIIIIIIIII|
+-- >                                              |
+-- >                                        Wildcard bit (bit 21)
 --
 -- If you change this representation, make sure to update 'transitionCodeUnit',
 -- 'wildcard', 'transitionState', 'transitionIsWildcard', 'newTransition' and
 -- 'newWildcardTransition' as well. Those functions form the interface used to
 -- construct and read transitions.
 type Transition = Word64
 
+-- TODO: Perhaps a word16 would be enough
```

Comment:
I think it might be too restrictive. Basically, we'd need to have fewer than 2^16 transitions overall, and while maybe not a common case, it's not unthinkable that we run into that at some point.

Reply:
I thought that as well and I think it's likely that this wouldn't do much for performance anyways.